ARROW-5681: [FlightRPC] Add Flight-specific error APIs

lihalite opened the pull request:
This adds a set of exceptions (in Java/Python) and status codes (in C++) for Flight, to convey Flight-specific issues and to let server implementations provide more semantic error messages to clients.

The error codes defined are a subset of the gRPC ones. In Flight/Java, you could always raise `io.grpc.StatusRuntimeException` yourself, but we want to abstract away from the gRPC APIs. 

There are not tests for conveying these statuses cross-language. That should be done as part of ARROW-5875.

Travis: https://travis-ci.com/lihalite/arrow/builds/118452703
AppVeyor: https://ci.appveyor.com/project/lihalite/arrow/builds/25849820

emkornfield commented on cpp/src/arrow/flight/internal.cc:
```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::ALREADY_EXISTS:
+      code = StatusCode::Invalid;
```
I believe I added an already exists error code.

lihalite replied:
Thanks for the pointer, I've changed that.

Do you think NOT_FOUND should also be promoted to the general status codes, as the counterpart?

emkornfield replied:
I think KeyError might be the equivalent.

lihalite replied:
Ah I see, thanks - updated.

emkornfield commented on cpp/src/arrow/flight/internal.cc:
```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+    case grpc::StatusCode::DEADLINE_EXCEEDED:
+      code = StatusCode::IOError;
+      message =
+          util::StringBuilder("gRPC returned deadline exceeded error, with message: ",
```
most if not all statuses can be constructed directly with:
Status::IOError("message ...", variable, "message ...").with_detail(details)

lihalite replied:
Thanks, that's much cleaner.

emkornfield commented on cpp/src/arrow/flight/internal.cc:
```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+                                    grpc_status.error_message());
+      break;
+    case grpc::StatusCode::INTERNAL:
+      code = StatusCode::IOError;
```
Unknown might be better?

lihalite replied:
Agreed, changed. I also added a missing branch for gRPC's UNKNOWN.

emkornfield commented on cpp/src/arrow/flight/internal.cc:
```diff
@@ -49,27 +50,153 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+  }
+
+  grpc::StatusCode grpc_code = grpc::StatusCode::UNKNOWN;
+  std::string message = arrow_status.message();
```
are you avoiding using Status.ToString() because it included Arrow's interpretation?

lihalite replied:
Yes, otherwise the error gets a bit repetitive.

lihalite commented:
Rebased/fixed conflicts.

emkornfield commented on cpp/src/arrow/flight/internal.cc:
```diff
@@ -49,27 +50,117 @@ Status FromGrpcStatus(const grpc::Status& grpc_status) {
+    case grpc::StatusCode::RESOURCE_EXHAUSTED:
+      return Status::Invalid("gRPC returned resource exhausted error, with message: ",
+                             grpc_status.error_message());
+    case grpc::StatusCode::FAILED_PRECONDITION:
```
FAILED_PRECONDITION, ABORTED, and RESOURCE_EXHAUSTED  errors might be good candidates to make top-level.  We should probably discuss on the mailing list.  But they have slightly different semantics and might be useful to build more general purpose machinery around handling them

emkornfield commented:
@lihalite this is kind of a big change.  If no one gets to it sooner I can try to have a more thorough review done sometime over the weekend but I have other competing priorities at the moment.

lihalite commented:
@emkornfield no worries, thanks for the comments so far. I'm not in a particular rush - over the next week is fine! I am trying to land this (+the follow up proposal on the mailing list) for 1.0.0, since it is a little painful to not really be able to tell what kind of error happened without inspecting repr/str.

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -37,6 +38,31 @@ import pyarrow.lib as lib
+        with gil:
+            message = frombytes(status.message())
+        if detail.get().code() == CFlightStatusInternal:
+            raise FlightInternalError(message)
```
I _think_ all the raises should be under the `with gil`, though perhaps Cython does it for us?

lihalite replied:
It "worked" before but I'd rather be explicit, thanks.

pitrou commented on cpp/src/arrow/flight/types.h:
```diff
@@ -51,6 +51,61 @@ class Uri;
+  ///
+  /// \return true if a \a FlightStatusDetail was unwrapped, false
+  /// otherwise
+  static bool UnwrapStatus(const arrow::Status& status,
```
Could simply return a `shared_ptr<FlightStatusDetail>` and use `nullptr` to signal that unwrapping failed.

lihalite replied:
Good point, thanks. I've gotten too used to making everything an out parameter by default...

lihalite commented:
Rebased, Travis build: https://travis-ci.com/lihalite/arrow/builds/119628842

pitrou commented on python/pyarrow/_flight.pyx:
```diff
@@ -73,6 +100,45 @@ class CertKeyPair(_CertKeyPair):
     """A TLS certificate and key for use in Flight."""
 
 
+cdef class FlightError(Exception):
```
Should we inherit from ArrowException here rather than below?

lihalite replied:
I had to do this dance to make Cython happy: ArrowException isn't cdef, so inheriting from it from a cdef class doesn't work.

pitrou replied:
Ok, that's fine with me.

lihalite commented:
@pitrou thanks! Yes, still waiting on a Java review.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/grpc/StatusUtils.java:
```diff
@@ -0,0 +1,192 @@
+      case NOT_FOUND:
+        return FlightStatusCode.INVALID_ARGUMENT;
+      case ALREADY_EXISTS:
+        return FlightStatusCode.INVALID_ARGUMENT;
```
per comment above these still seem like INVALID_ARGUMENT might be too broad a catch-all.

lihalite replied:
I added these two codes to the Flight codes, to mirror what was done in C++/Python.

emkornfield commented on java/flight/src/test/java/org/apache/arrow/flight/auth/TestBasicAuth.java:
```diff
@@ -79,20 +78,23 @@ public void asyncCall() {
-    assertThrows(StatusRuntimeException.class, () -> {
-      client.listFlights(Criteria.ALL);
-    }, PERMISSION_DENIED);
+    ex = assertThrows(FlightRuntimeException.class, () -> {
```
it might pay to making a "testing" utility class that combines these to lines together.

lihalite replied:
Good idea, thanks.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/FlightStatusCode.java:
```diff
@@ -0,0 +1,42 @@
+   * Create a blank {@link CallStatus} with this code.
+   */
+  public CallStatus toStatus() {
+    return new CallStatus(this, null, null);
```
please document what the null parameter represent.

lihalite replied:
I ended up creating the single-parameter constructor as suggested below.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Objects;
+
+/**
+ * The result of a Flight RPC.
```
can you expand on this comment.  In particular, it is not immediately obvious why this can't just be an exception class (i.e. why do we need to represent OK)?  It seems like this might have do with the flight state machine if I understand the code correctly.

lihalite replied:
I've mostly followed the gRPC API in this regard, which has both checked and unchecked versions of the exception that use a common status object. We don't need to represent OK and don't use the status, so I've removed it. I did expand the comment a bit.

emkornfield commented on java/flight/src/main/java/org/apache/arrow/flight/CallStatus.java:
```diff
@@ -0,0 +1,95 @@
+   * @param cause An exception that resulted in this status (or null).
+   * @param description A description of the status (or null).
+   */
+  public CallStatus(FlightStatusCode code, Throwable cause, String description) {
```
consider making a construct that just take code, so you don't have to make nullable parameters for constructing from the the enum.

emkornfield commented:
Overall, seems ok to me on the java side, left a few comment.  I'm not entirely sure everything should be translated to a runtime exception but we can revisit this if it becomes an issue later.  I still think it might be discussing exposing a few more Status codes at the top-level instead of conflating them, but this is also something we can evolve as we more usage on flight (hopefully we get  feedback before we declare flight stable otherwise the migration might be a little painful).

lihalite commented:
Thanks for the review! I addressed your feedback and rebased. The Java API is modeled off of the gRPC API, but that isn't necessarily a requirement (e.g we could use checked exceptions if we feel that makes more sense).

lihalite commented:
Rebased again, personal Travis: https://travis-ci.com/lihalite/arrow/builds/120745269

pitrou commented:
Thanks @lihalite !

lihalite commented:
Thanks all for looking over this! On to the next task :)
